java/openmldb-batch/src/main/scala/com/_4paradigm/openmldb/batch/nodes/JoinPlan.scala

```diff
@@ -119,76 +118,67 @@ object JoinPlan {
 
       // Handle the duplicated column names to get Spark Column by index
       val allColumns = new mutable.ArrayBuffer[Column]()
-      for (i <- 0 until leftDf.schema.size) {
+      for (i <- leftDf.schema.indices) {
         if (i != indexColIdx) {
           allColumns += SparkColumnUtil.getColumnFromIndex(leftDf, i)
         }
       }
-      for (i <- 0 until rightDf.schema.size) {
+      for (i <- rightDf.schema.indices) {
         allColumns += SparkColumnUtil.getColumnFromIndex(rightDf, i)
       }
 
-      val allColWrap = functions.struct(allColumns:_*)
+      val allColWrap = functions.struct(allColumns: _*)
       joinConditions += functions.callUDF(regName, allColWrap)
     }
 
     if (joinConditions.isEmpty) {
       throw new HybridSeException("No join conditions specified")
     }
 
-    val joined = leftDf.join(rightDf, joinConditions.reduce(_ && _),  "left")
+    val joined = leftDf.join(rightDf, joinConditions.reduce(_ && _), "left")
 
     val result = if (joinType == JoinType.kJoinTypeLeft) {
       joined
     } else if (joinType == JoinType.kJoinTypeLast) {
       // Resolve order by column index
       if (hasOrderby) {
         val orderExpr = node.join.right_sort.orders.GetOrderExpression(0)
-        val planLeftSize = node.GetProducer(0).GetOutputSchema().size()
         // Get the time column index from right table
         val timeColIdx = SparkColumnUtil.resolveOrderColumnIndex(orderExpr, node.GetProducer(1))
         assert(timeColIdx >= 0)
-
         val timeIdxInJoined = timeColIdx + leftDf.schema.size
-        val timeColType = rightDf.schema(timeColIdx).dataType
 
         val isAsc = node.join.right_sort.is_asc
 
-        // Disable scalastyle for importing Spark implicits
-        // scalastyle:off
-        import spark.implicits._
-        // scalastyle:on
-        val distinct = joined
-          .groupByKey {
-            row => row.getLong(indexColIdx)
-          }
-          .mapGroups {
-            case (_, iter) =>
-              val timeExtractor = SparkRowUtil.createOrderKeyExtractor(
-                timeIdxInJoined, timeColType, nullable=false)
-
-              if (isAsc) {
-                iter.maxBy(row => {
-                  if (row.isNullAt(timeIdxInJoined)) {
-                    Long.MinValue
-                  } else {
-                    timeExtractor.apply(row)
-                  }
-                })
+        val indexCol = SparkColumnUtil.getColumnFromIndex(joined, indexColIdx)
+
+        val distinctRdd = joined.repartition(indexCol).rdd.map({
+          row => (row.getLong(indexColIdx), row)
+        }).reduceByKey({
+          (row1, row2) =>
+            val dataType = joined.schema.fields(timeIdxInJoined).dataType
+            val data1 = SparkRowUtil.getLongFromIndex(timeIdxInJoined, dataType, row1)
+            val data2 = SparkRowUtil.getLongFromIndex(timeIdxInJoined, dataType, row2)
+            if (isAsc) {
+              if (data1 > data2) {
```

Comment:
In scala, single line for simple if-else is more readable. Could be changed to this.

`if (rowTimeValue1 > rowTimeValue2) row1 else row2`.